repository-azure/src/main/java/io/aiven/elasticsearch/repositories/azure/AzureClient.java

```diff
@@ -0,0 +1,86 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.elasticsearch.repositories.azure;
+
+import java.util.Objects;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import com.azure.core.http.okhttp.OkHttpAsyncHttpClientBuilder;
+import com.azure.core.http.policy.UserAgentPolicy;
+import com.azure.storage.blob.BlobServiceClient;
+import com.azure.storage.blob.BlobServiceClientBuilder;
+import com.azure.storage.common.policy.RequestRetryOptions;
+import com.azure.storage.common.policy.RetryPolicyType;
+import okhttp3.Dispatcher;
+
+class AzureClient implements AutoCloseable {
+
+    static final String HTTP_USER_AGENT = "aiven-azure-repository";
+
+    private final BlobServiceClient blobServiceClient;
+
+    /**
+     * Thread pool for Azure SDK HTTP client, must be closed explicitly
+     */
+    private final ExecutorService httpPoolExecutorService;
+
+    private AzureClient(final BlobServiceClient blobServiceClient, final ExecutorService httpPoolExecutorService) {
+        this.blobServiceClient = blobServiceClient;
+        this.httpPoolExecutorService = httpPoolExecutorService;
+    }
+
+    public BlobServiceClient blobServiceClient() {
+        return blobServiceClient;
+    }
+
+    public static AzureClient create(final AzureStorageSettings azureStorageSettings) {
+        final var httpThreadPoolSettings = azureStorageSettings.httpThreadPoolSettings();
+        final var httpClientExecutor = new ThreadPoolExecutor(
+                httpThreadPoolSettings.minThreads(), httpThreadPoolSettings.maxThreads(),
+                httpThreadPoolSettings.keepAlive(), TimeUnit.SECONDS,
+                new SynchronousQueue<>()
+        );
+        final var blobServiceClient = new BlobServiceClientBuilder()
+                .connectionString(azureStorageSettings.azureConnectionString())
+                .retryOptions(
+                        new RequestRetryOptions(
+                                RetryPolicyType.EXPONENTIAL, azureStorageSettings.maxRetries(),
+                                null, null, null, null))
+                .addPolicy(new UserAgentPolicy(HTTP_USER_AGENT))
+                .httpClient(
+                        new OkHttpAsyncHttpClientBuilder()
+                                .dispatcher(new Dispatcher(httpClientExecutor))
+                                .build())
+                .buildClient();
+        return new AzureClient(blobServiceClient, httpClientExecutor);
+    }
+
+    @Override
+    public void close() {
+        httpPoolExecutorService.shutdown();
+        try {
+            if (!httpPoolExecutorService.awaitTermination(800L, TimeUnit.MILLISECONDS)) {
+                httpPoolExecutorService.shutdownNow();
```

Comment:
What do you think about waiting a little bit (call awaitTermination once again) after shutdownNow request, just to make sure that everything is OK?

Reply:
It can try 3 times and after that calls `shutdownNow`. Reasonable?

Reply:
yep